tensorflow/python/training/optimizer.py

```diff
@@ -571,6 +571,7 @@ def _get_or_make_slot(self, var, val, slot_name, op_name):
     Returns:
       A `Variable` object.
     """
+    op_name += '/' + slot_name
```

Comment:
Yeah, there may be something hard to understand.
## The reason is:
Some optimizers have to create more than one list of shared variables. As the master's implementation, it use `ops.name_scope(name) as scope` to distinguish the creations of different variables' lists. It will add a suffix to the scope name if a variable with such name already exists. So it will give two lists of variables(call 'm' list and 'v' list) in Adam like below:
```
embedding_attention_seq2seq/RNN/EmbeddingWrapper/embedding/Adam:0 (50000, 256)
embedding_attention_seq2seq/RNN/EmbeddingWrapper/embedding/Adam_1:0 (50000, 256)
```
And in my implementation, it doesn't allow the repeated creation of variable with the same scope name. So the Adam optimizer will fail to create its second list of variables which is called 'v' in Adam.
```python
# Adam will call _zeros_slot twice
    for v in var_list:
      self._zeros_slot(v, "m", self._name)
      self._zeros_slot(v, "v", self._name)
```
To solve this issue, it need to use different scope names between 'm' list and 'v' list in Adam.
Fortunately the slot_name in optimizer is not used now, so I think `op_name + '/' + slot_name` is suitable to use as different scope name to deal with this issue.
```python
def _zeros_slot(self, var, slot_name, op_name)
    # add here
    op_name += '/' + slot_name
    named_slots = self._slot_dict(slot_name)
    if var not in named_slots:
      named_slots[var] = slot_creator.create_zeros_slot(var, op_name)
    return named_slots[var]
```
And meanwhile using op_name with a number as the suffix may be confused, using `op_name + slot_name` will be clearer.
```
embedding_attention_seq2seq/RNN/EmbeddingWrapper/embedding/Adam/m:0 (50000, 256)
embedding_attention_seq2seq/RNN/EmbeddingWrapper/embedding/Adam/v:0 (50000, 256)
```